src/SAD/Parser/Token.hs

```diff
@@ -72,60 +79,69 @@ isProperToken EOF{} = True
+    posToken texState pos whitespaceBefore s | not (Text.null lexeme) = tok:toks
       where
-        (lexem, rest) = Text.span isLexem s
-        tok  = makeToken lexem pos whitespaceBefore mode
-        toks = posToken (advanceAlong pos lexem) NoWhiteSpaceBefore mode rest
-    posToken pos _ mode s | not (Text.null white) = toks
+        (lexeme, rest) = Text.span isLexeme s
+        tok  = makeToken lexeme pos whitespaceBefore
+        toks = posToken texState (advanceAlong pos lexeme) NoWhiteSpaceBefore rest
+    
+    -- Process whitespace.
+    posToken texState pos _ s | not (Text.null white) = toks
       where
         (white, rest) = Text.span isSpace s
-        toks = posToken (advanceAlong pos white) WhiteSpaceBefore mode rest
-    posToken pos whitespaceBefore mode s = case Text.uncons s of
+        toks = posToken texState (advanceAlong pos white) WhiteSpaceBefore rest
+    
+    -- Process non-alphanumeric symbol or EOF.
+    posToken texState pos whitespaceBefore s = case Text.uncons s of
       Nothing -> [EOF pos]
       Just ('\\', rest) -> tok:toks
         where
           (name, rest') = Text.span isAlpha rest
-          cmd = (Text.cons '\\' name)
-          tok = makeToken cmd pos whitespaceBefore mode
-          toks = posToken (advanceAlong pos cmd) WhiteSpaceBefore mode rest'
-      Just ('$', rest) -> toks
-        where
-          toks = posToken (advancePos pos '$') NoWhiteSpaceBefore (switchMode mode) rest
-          switchMode :: TokenMode -> TokenMode
-          switchMode TextMode = MathMode
-          switchMode MathMode = TextMode
-      Just ('#', _) -> tok:toks
-        where
-          (comment, rest) = Text.break (== '\n') s
-          tok  = makeToken comment pos Comment mode
-          toks = posToken (advanceAlong pos comment) whitespaceBefore mode rest
-      Just ('%', _) -> tok:toks
+          cmd = Text.cons '\\' name
+          tok = makeToken cmd pos whitespaceBefore
+          toks = posToken texState (advanceAlong pos cmd) WhiteSpaceBefore rest'
+      Just (c, _) | if usingTexParser texState then c == '%' else c `elem` ['%','#'] -> tok:toks
         where
           (comment, rest) = Text.break (== '\n') s
-          tok  = makeToken comment pos Comment mode
-          toks = posToken (advanceAlong pos comment) whitespaceBefore mode rest
+          tok  = makeToken comment pos Comment
+          toks = posToken texState (advanceAlong pos comment) whitespaceBefore rest
       Just (c, cs) -> tok:toks
         where
-          tok  = makeToken (Text.singleton c) pos whitespaceBefore mode
-          toks = posToken (advancePos pos c) NoWhiteSpaceBefore mode cs
+          tok  = makeToken (Text.singleton c) pos whitespaceBefore
+          toks = posToken texState (advancePos pos c) NoWhiteSpaceBefore cs
 
 
-isLexem :: Char -> Bool
-isLexem c = (isAscii c && isAlphaNum c) || c == '_'
+isLexeme :: Char -> Bool
+isLexeme c = isAscii c && isAlphaNum c
```

Comment:
I think '_' is quite useful in normal forthel texts. Maybe this could also depend on `texState`?

Reply:
I added `_` again now.